File: saga-architecture-net/PaymentService/Application/PaymentService.cs
using Saga.PaymentService.Common;
using Saga.PaymentService.Domain;

namespace Saga.PaymentService.Application;

public interface IBalanceRepository
{
    Task<UserBalance?> GetBalance(int userId);
    Task UpdateBalance(UserBalance balance);
}

public interface IPaymentTransactionRepository
{
    Task SaveTransaction(PaymentTransaction transaction);
    Task<PaymentTransaction?> GetTransaction(Guid orderId);
    Task UpdateTransaction(PaymentTransaction transaction);
}

public class PaymentService(IBalanceRepository balanceRepo, IPaymentTransactionRepository txnRepo, IKafkaProducer producer)
{
    public async Task<PaymentEvent> ProcessPayment(PaymentRequestDTO request)
    {
        var balance = await balanceRepo.GetBalance(request.UserId);
        var txn = new PaymentTransaction { OrderId = request.OrderId, UserId = request.UserId, Amount = request.Amount };

        if (balance != null && balance.Balance >= request.Amount)
        {
            balance.Balance -= request.Amount;
            await balanceRepo.UpdateBalance(balance);
            txn.Status = "APPROVED";
            await txnRepo.SaveTransaction(txn);
            var evt = new PaymentEvent(request, PaymentStatus.PAYMENT_COMPLETED);
            await producer.ProduceAsync("payment-events", evt);
            return evt;
        }
        else
        {
            txn.Status = "REJECTED";
            await txnRepo.SaveTransaction(txn);
            var evt = new PaymentEvent(request, PaymentStatus.PAYMENT_FAILED);
            await producer.ProduceAsync("payment-events", evt);
            return evt;
        }
    }

    public async Task RefundPayment(PaymentRequestDTO request)
    {
        var txn = await txnRepo.GetTransaction(request.OrderId);
        if (txn != null && txn.Status == "APPROVED")
        {
            var balance = await balanceRepo.GetBalance(request.UserId);
            if (balance != null)
            {
                balance.Balance += txn.Amount;
                await balanceRepo.UpdateBalance(balance);
                txn.Status = "REFUNDED";
                await txnRepo.UpdateTransaction(txn);
            }
        }
    }

    public async Task<UserBalance?> GetUserBalance(int userId) => await balanceRepo.GetBalance(userId);
}

public interface IKafkaProducer
{
    Task ProduceAsync<T>(string topic, T message);
}

File: saga-architecture-java/payment-service/src/main/java/com/example/saga/payment/service/PaymentService.java
package com.example.saga.payment.service;

import com.example.saga.common.PaymentRequestDTO;
import com.example.saga.common.PaymentEvent;
import com.example.saga.common.PaymentStatus;
import com.example.saga.payment.entity.PaymentTransaction;
import com.example.saga.payment.entity.UserBalance;
import com.example.saga.payment.repository.PaymentTransactionRepository;
import com.example.saga.payment.repository.UserBalanceRepository;
import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
@Slf4j
@RequiredArgsConstructor
public class PaymentService {

    private final UserBalanceRepository balanceRepository;
    private final PaymentTransactionRepository transactionRepository;

    @Transactional
    public PaymentEvent processPayment(PaymentRequestDTO paymentRequest) {
        log.info("Processing payment for order: {}", paymentRequest.orderId());

        PaymentTransaction transaction = new PaymentTransaction();
        transaction.setOrderId(paymentRequest.orderId());
        transaction.setUserId(paymentRequest.userId());
        transaction.setAmount(paymentRequest.amount());
        transaction.setProductId(paymentRequest.productId());

        return balanceRepository.findById(paymentRequest.userId())
                .filter(ub -> ub.getBalance() >= paymentRequest.amount())
                .map(ub -> {
                    ub.setBalance(ub.getBalance() - paymentRequest.amount());
                    balanceRepository.save(ub);
                    transaction.setStatus("APPROVED");
                    transactionRepository.save(transaction);
                    return new PaymentEvent(paymentRequest, PaymentStatus.PAYMENT_COMPLETED);
                })
                .orElseGet(() -> {
                    transaction.setStatus("REJECTED");
                    transactionRepository.save(transaction);
                    return new PaymentEvent(paymentRequest, PaymentStatus.PAYMENT_FAILED);
                });
    }

    @Transactional
    public void refundPayment(PaymentRequestDTO paymentRequest) {
        log.info("Refunding payment for order: {}", paymentRequest.orderId());

        // Find successful transaction to refund
        transactionRepository.findById(paymentRequest.orderId()).ifPresent(txn -> {
            balanceRepository.findById(paymentRequest.userId()).ifPresent(ub -> {
                ub.setBalance(ub.getBalance() + txn.getAmount());
                balanceRepository.save(ub);
                log.info("Refunded {} to user {}", txn.getAmount(), paymentRequest.userId());
            });
            txn.setStatus("REFUNDED");
            transactionRepository.save(txn);
        });
    }

    @Transactional(readOnly = true)
    public UserBalance getUserBalance(Integer userId) {
        return balanceRepository.findById(userId).orElse(null);
    }
}
